index.html

```diff
@@ -2175,6 +2231,130 @@ <h3>Verify Proof Sets and Chains</h3>
+
+        <p>
+The purpose of the algorithm in this section is to ensure that a consuming
+application has explicitly approved of the types, and therefore the semantics,
+of input documents that it will process. Not checking JSON-LD context values
+against known good values can lead to security vulnerabilities, due to variance
+in the semantics that they convey. The context validation algorithm is one
+that applications validating a [=conforming secured document=] MUST run after
+running the algorithm in Section [[[#verify-proof]]] or Section
+[[[#verify-proof-sets-and-chains]]]. This algorithm takes a document ([=map=]
+|inputDocument|), a set of approved JSON-LD Contexts ([=map=]
+|approvedContexts|), and a boolean to recompact when unknown contexts are
+detected ([=boolean=] |recompact|) and returns a [=map=] that contains the
+following:
+        </p>
+
+        <ul>
+          <li>
+a status ([=boolean=] |status|)
+          </li>
+          <li>
+a validated document ([=map=] |document|) if no errors occurred
+          </li>
+          <li>
+zero or more warnings ([=list=] of [=ProblemDetails=] |warnings|)
+          </li>
+          <li>
+zero or more errors ([=list=] of [=ProblemDetails=] |errors|)
+          </li>
+        </ul>
+
+        <p>
+The context validation algorithm is as follows:
+        </p>
+
+        <ol class="algorithm">
+          <li>
+Set |result|.|status| to `false`, |result|.|warnings| to an empty list,
+|result|.|errors| to an empty list, |compactionContext| to an empty list,
+and clone |inputDocument| to |result|.|document|.
+          </li>
+          <li>
+For each subtree in |result|.|document| that contains a `@context` property:
+            <ol class="algorithm">
+              <li>
+For each value (|contextValue|) associated with the property:
+                <ol class="algorithm">
+                  <li>
+If |recompact| is `true` and the |contextValue| exists in |approvedContexts|,
+add the |contextValue| to the |compactionContext|.
+                  </li>
+                  <li>
+Otherwise, ensure that the contents of the |contextValue|, which is a JSON-LD
+Context file, match a known good value or cryptographic hash. If the check
+fails, add at least one error to |result|.|errors|.
+                  </li>
+                </ol>
+              </li>
+              <li>
+If |recompact| is `true`, set the subtree in |result|.|document| to the result
+of running the <a data-cite="JSON-LD11-API#compaction-algorithm">
+JSON-LD Compaction Algorithm</a> with the |inputDocument| and
+|compactionContext| as inputs. If the compaction fails, add at least one error
+to |result|.|errors|.
+              </li>
+            </ol>
+          </li>
+          <li>
+If |result|.|errors| is empty, set |result|.|status| to `true`. Otherwise, set
+|result|.|status| to `false`, and remove the |document| property from |result|.
+          </li>
+          <li>
+Return the value of |result|.
+          </li>
+        </ol>
```

Comment:
I think this algorithm is more complex than it needs to be, is unlikely to be used in very complex use cases anyway as JSON schema checks + compaction will be used more easily leveraged for those, and it won't work properly anyway because of the changing semantics in each nested body (i.e., the "approvedContexts" would need to match the nesting too, which is much better served by and understood through the use of JSON schema).

I also think that this algorithm can also result (when `recompact=false`) in a document that cannot be consumed because it contained no approved contexts, but somehow the cryptographic hashes matched known good values (this could only happen with a URL scheme that would allow hashes as a component of the context URL). But this sort of document still cannot be consumed by the application because the contexts aren't approved.

I recommend simplifying this algorithm for the less complex cases that have a particular top-level context in the document and make `approvedContexts` simpler too: it can just become become an expected context value (a single context value or a list of context values in a particular order) as `expectedContext`. If the document does not have exactly the `expectedContext`, it should be recompacted to `expectedContext` or rejected (depending on the `recompact` flag).

I may be missing one of the aims here, but this approach seems far simpler to me:

```suggestion
          <li>
Let |contextValue| be the value of the `@context` property of |result|.|document|,
which might be undefined.
          </li>
          <li>
If |contextValue| does not deeply equal |expectedContext| or if any
any subtree in |result|.|document| contains an `@context` property:
            <ol class="algorithm">
              <li>
If |recompact| is `true` set |result|.|document| to the result
of running the <a data-cite="JSON-LD11-API#compaction-algorithm">
JSON-LD Compaction Algorithm</a> with the |inputDocument| and
|expectedContext| as inputs. If the compaction fails, add at least one error
to |result|.|errors|.
              </li>
              <li>
Otherwise, add at least one error to |result|.|errors|.
              </li>
            </ol>
          </li>
          <li>
If |result|.|errors| is empty, set |result|.|status| to `true`. Otherwise, set
|result|.|status| to `false`, and remove the |document| property from |result|.
          </li>
          <li>
Return the value of |result|.
          </li>
        </ol>
```

Reply:
The above could possibly be tweaked for readability a bit more, but wanted to get this suggestion out for review.